From 2b3e1c379bf9b9404452b7aa1a66a6ce6d1301e3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Mattias=20Walstr=C3=B6m?= <lazzer@gmail.com>
Date: Fri, 17 May 2024 13:33:53 +0200
Subject: [PATCH] Add support for system authentication as fallback

First check if the user is configurured in ietf-netconf-server YANG,
if found in YANG go with the user found. Else, if using shadow or pam,
just continue, try to auth anyway. PAM will block the login later if
incorrect.
---
 src/session_server_ssh.c | 17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/src/session_server_ssh.c b/src/session_server_ssh.c
index c8386bc..86d5bb1 100644
--- a/src/session_server_ssh.c
+++ b/src/session_server_ssh.c
@@ -1415,6 +1415,7 @@ nc_session_ssh_msg(struct nc_session *session, struct nc_server_ssh_opts *opts,
      * process known messages
      */
     if (type == SSH_REQUEST_AUTH) {
+        int username_allocated = 0;
         if (session->flags & NC_SESSION_SSH_AUTHENTICATED) {
             ERR(session, "User \"%s\" authenticated, but requested another authentication.", session->username);
             ssh_message_reply_default(msg);
@@ -1439,6 +1440,15 @@ nc_session_ssh_msg(struct nc_session *session, struct nc_server_ssh_opts *opts,
             }
         }
 
+#if defined(HAVE_LIBPAM) || defined(HAVE_SHADOW)
+	if (!auth_client) {
+		username_allocated = 1;
+		auth_client = malloc(sizeof(*auth_client)); /* Get freed when closing session. */
+		memset(auth_client, 0, sizeof(*auth_client));
+		auth_client->username = strdup(username);
+		auth_client->kb_int_enabled = 1;
+	}
+#else
         if (!auth_client) {
             if (opts->referenced_endpt_name) {
                 /* client not known by the endpt, but it references another one so try it */
@@ -1457,9 +1467,12 @@ nc_session_ssh_msg(struct nc_session *session, struct nc_server_ssh_opts *opts,
             ssh_message_reply_default(msg);
             return 0;
         }
-
+#endif
         if (!session->username) {
-            session->username = strdup(username);
+            if (username_allocated)
+               session->username = auth_client->username;
+            else
+               session->username = strdup(username);
 
             /* configure and count accepted auth methods */
             if (auth_client->store == NC_STORE_LOCAL) {
-- 
2.34.1

